src/placeos-frontend-loader/api/remote/githubremote.cr

```diff
@@ -0,0 +1,168 @@
+require "hash_file"
+require "./remote"
+require "octokit"
+
+module PlaceOS::FrontendLoader
+  class GitHubRemote < Remote
+    def initialize
+    end
+
+    private alias Remote = PlaceOS::FrontendLoader::Remote
+
+    TAR_NAME = "temp.tar.gz"
+
+    @github_client = Octokit.client(GIT_USER, GIT_PASS)
+
+    # Returns the branches for a given repo
+    def branches(repo : String) : Hash(String, String)
+      get_branches = @github_client.branches(repo)
+      branches = Hash(String, String).new
+      get_branches.fetch_all.each do |branch|
+        next if branch =~ /HEAD/
+        branches[branch.name] = branch.commit.sha
+      end
+      branches
+    end
+
+    # Returns the commits for a given repo on specified branch
+    def commits(repo : String, branch : String) : Array(Remote::Commit)
+      get_commits = @github_client.commits(repo, branch)
+      commits = Array(Remote::Commit).new
+      get_commits.fetch_all.each do |comm|
+        commit = Commit.new(
+          commit: comm.sha,
+          date: comm.commit.author.date,
+          author: comm.commit.author.name,
+          subject: comm.commit.message,
+        )
+        commits << commit
+      end
+      commits
+    end
+
+    # Returns the release tags for a given repo
+    def releases(repo : String) : Array(String)
+      get_tags = @github_client.tags(repo)
+      tags = Array(String).new
+      get_tags.fetch_all.each { |tag| tags << tag.name }
+      tags
+    end
+
+    def url(repo_name : String) : String
+      "https://github.com/#{repo_name}"
+    end
+
+    def download(
+      ref : Remote::Reference,
+      branch : String? = "master",
+      hash : String? = "HEAD",
+      tag : String? = "latest",
+      path : String = "./"
+    )
+      repository_uri = url(ref.repo_name)
+      repository_folder_name = path.split("/").last
+
+      if hash == "HEAD" || hash.nil?
+        if tag != "latest"
+          hash = get_hash_by_tag(repository_uri, tag)
+        elsif branch != "master"
+          hash = get_hash_by_branch(repository_uri, branch)
+        else
+          hash = get_hash_head(repository_uri)
+        end
+        hash = hash.not_nil!
+      end
```

Comment:
May be a bit too much indirection, but I think putting this inside the ref instance would be good.

```crystal
struct Reference
  def reference(remote) : String
    # Logic based off calls to the remote
  end
end
```